### PR TITLE
#1032: Migrates of PFADD, PFCOUNT, PFMERGE to Multi-threaded architecture.

### LINKED ISSUE #1032: Command Migration: ('PFADD', 'PFCOUNT', 'PFMERGE')
This issue tracks the migration of the mentioned commands - ('PFADD', 'PFCOUNT', 'PFMERGE') to make them compatible across the three protocols supported by the Dice database: RESP, HTTP, and WebSocket.
DiceDB now supports more than one protocols (Resp/http/websocket). We don't want eval operation of each command to be strictly bound with any of the protocols. Currently eval function return statements of each command is bounded to the RESP protocol. 

*The migration is required because:*

- The current implementation is specific to the RESP protocol
- Evaluation functions of each command must be refactored to ensure protocol-agnostic behavior.
- Evaluation functions should return raw responses instead of protocol specific responses.

The goal is to make the command logic protocol-independent, allowing all three protocols to call the same core functionality seamlessly.

### Requirements

- Refactor the evaluation function of `[Command Name]` to be generic.
- Ensure the evaluation function does not include protocol-specific logic.

### Migration Steps

1. **Analyze Current Implementation**
   - Review the current codebase to understand how the command logic is implemented.

2. **Refactor Return Logic**
   - Create a new function with the function definition as `evalXXX(args []string, store *dstore.Store) *EvalResponse` under the file `/internal/eval/store_eval.go`
   - Analyse return statements of the eval and modify them to send raw types without encoding
   - Use errors in the return statements from `/errors/migrated_errors.go` file
   - Use perdefined responses from `/internal/clientio/resp.go` file as RespType variables

3. **Command/Worker specific Changes**
   - Make `IsMigrated` flag to true in the commands information under `/internal/eval/commands.go` file.
   - Use the newly written eval function against the `NewEval` parameter in the same command structure.
   - Delete the old eval function from the `/internal/eval/eval.go` file.
   - Add the migrated command to the `/internal/worker/CommandsMeta` map and make it's type as `SingleShard`

4. **Update Unit Tests**
   - Refactor existing unit tests to accommodate the new implementation.
   - Add new unit tests if necessary to cover all possible cases.

5. **Integration Tests**
   - Run all integration tests to ensure successful migration.
   - Ensure that each protocol (RESP, HTTP, WebSocket) works correctly after migration.

### Checklist

- [ ] Migrated the evalXXX function with the latest definition
- [ ] Update or add unit tests for the new implementation.
- [ ] All unit tests pass successfully.
- [ ] Ensure all integration tests pass successfully.

### Additional Notes

- Describe any edge cases that need to be handled in the generic function.
- Mention any protocol-specific optimizations required in wrappers.
  
If there are any questions or concerns about this migration, please mention them here.

### Related Issues/PRs

- Sample implementation for the Get, Set, GetSet and SetEx can be found in [this Pull Request.](https://github.com/DiceDB/dice/pull/766)

### PR DESCRIPTION
**Pull request by surya0180**
This PR closes #1032.

Here are things that are done

 - [x] Migrated the evalXXX function with the latest definition
 - [x] Update or add unit tests for the new implementation.
 - [x] All unit tests pass successfully.
 - [x] Ensure all integration tests pass successfully.
 - [x] Ensure integration tests are added for the migrated command on multi-threaded resp server.

### CODE REVIEW

File: internal/eval/eval_test.go [hunk 1]
```diff
@@ -44,79 +44,79 @@ func setupTest(store *dstore.Store) *dstore.Store {
...
-	testEvalPersist(t, store)
-	testEvalEXPIRE(t, store)
-	testEvalEXPIRETIME(t, store)
-	testEvalEXPIREAT(t, store)
-	testEvalDbsize(t, store)
-	testEvalGETSET(t, store)
-	testEvalHSET(t, store)
-	testEvalHMSET(t, store)
-	testEvalHKEYS(t, store)
+	// testEvalMSET(t, store)
```
Comment (benjaminmishra) [inline internal/eval/eval_test.go:47]: Should we have these test functions commented out ?

File: internal/eval/eval_test.go [hunk 2]
```diff
@@ -2103,12 +2182,181 @@ func testEvalPFCOUNT(t *testing.T, store *dstore.Store) {
...
+			input: []string{"EXISTING_KEY"},
+			migratedOutput: EvalResponse{
+				Result: uint64(1),
+				Error:  nil,
+			},
 		},
 	}
 
-	runEvalTests(t, tests, evalPFCOUNT, store)
+	for _, tt := range tests {
```
Comment (benjaminmishra) [inline internal/eval/eval_test.go:2193]: This part seems to be same for all 3 tests. May be combine them into one function and call that function here. Or you can also take a look at `runEvalTests` or `runMigratedEvalTests` function and see if they can be reused here ?

**Review by benjaminmishra — CHANGES_REQUESTED** [review]

File: internal/eval/eval_test.go (hunk 2 above)
Comment (surya0180) [reply internal/eval/eval_test.go:2193]: I could have written it as a single function and reuse it, but again, there are so many people working on migrations and if there are multiple of those functions with different names, they would all be redundant. Since you are suggesting to use `runMigratedEvalTests` I will go with that. Thanks for the suggestion

File: internal/eval/eval_test.go (hunk 1 above)
Comment (surya0180) [reply internal/eval/eval_test.go:47]: Oh shoot, I thought to uncomment them before making a commit. I made a mistake here. Let me quickly correct it.

**Comment by surya0180** [comment]
@benjaminmishra I have made the changes as requested. Please review them and let me know if everything looks fine

File: internal/eval/store_eval.go [hunk 3]
```diff
@@ -448,3 +449,154 @@ func evalZRANGE(args []string, store *dstore.Store) *EvalResponse {
...
+			}
+		}
+	}
+
+	// Save the mergedHll
+	obj = store.NewObj(mergedHll, -1, object.ObjTypeString, object.ObjEncodingRaw)
+	store.Put(destKey, obj)
+
+	return &EvalResponse{
+		Result: "OK",
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:599]: Instead of sending a response as "OK" you can return back `clientio.OK`

**Comment by soumya-codes** [comment]
@surya0180 lets ensure that the corresponding integration tests are migrated as well.

**Review by AshwinKul28 — CHANGES_REQUESTED** [review]
Everything looks fantastic @surya0180 . Just left one minor comment.

**Review by benjaminmishra — APPROVED** [review]
Looks good

**Comment by AshwinKul28** [comment]
@surya0180 Please add respective RESP integration tests. Thanks

File: internal/eval/store_eval.go (hunk 3 above)
Comment (surya0180) [reply internal/eval/store_eval.go:599]: Ok, I will make this change

**Comment by surya0180** [comment]
> @surya0180 lets ensure that the corresponding integration tests are migrated as well. cc @AshwinKul28

Is this what @soumya-codes is referring to?

File: internal/eval/store_eval.go (hunk 3 above)
Comment (surya0180) [reply internal/eval/store_eval.go]: I have made this change. Also added the Integration Tests for RESP Protocol

**Review by JyotinderSingh — APPROVED** [review]
Thanks for making this massive effort @surya0180! Thanks for reviewing @AshwinKul28, @soumya-codes, @benjaminmishra
